mono/utils/mono-threads.c

```diff
@@ -1835,3 +1911,181 @@ mono_thread_info_get_tools_data (void)
 
 	return info ? info->tools_data : NULL;
 }
+
+typedef struct {
+	MonoLockFreeQueueNode node;
+
+	// For cleanup metadata
+	MonoUtilityThread *thread;
+
+	// For synch calls
+	gboolean *finished;
+	MonoSemType *response_sem;
+
+	// Variably-sized, size is thread->payload_size
+	gpointer payload [MONO_ZERO_LEN_ARRAY]; 
+} UtilityThreadQueueEntry;
+
+static void
+free_queue_entry (gpointer p)
+{
+	UtilityThreadQueueEntry *util = (UtilityThreadQueueEntry *) p;
+	mono_lock_free_free (p, util->thread->message_block_size);
+}
+
+static gboolean
+utility_thread_handle_inbox (MonoUtilityThread *thread)
+{
+	UtilityThreadQueueEntry *entry = (UtilityThreadQueueEntry *) mono_lock_free_queue_dequeue (&thread->work_queue);
+	if (!entry)
+		return FALSE;
+	
+	thread->callbacks.command (thread->state_ptr, &entry->payload);
+	if (entry->response_sem) {
+		*entry->finished = TRUE;
+		mono_os_sem_post (entry->response_sem);
+	}
+
+	mono_thread_hazardous_try_free (entry, free_queue_entry);
+
+	return TRUE;
+}
+
+static void *
+utility_thread (void *arg)
+{
+	mono_thread_info_wait_inited ();
+	mono_thread_info_attach ();
+
+	MonoUtilityThread *thread = (MonoUtilityThread *) arg;
+	thread->callbacks.init (&thread->state_ptr);
+
+	while (mono_atomic_load_i32 (&thread->run_thread)) {
+		MONO_ENTER_GC_SAFE;
+		mono_os_sem_timedwait (&thread->work_queue_sem, 1000, MONO_SEM_FLAGS_NONE);
+		MONO_EXIT_GC_SAFE;
+		utility_thread_handle_inbox (thread);
+	}
```

Comment:
The profiler also needs a callback here. Either it has to be a separate callback or the command callback needs a Boolean parameter indicating whether the inbox is being drained on shutdown or if it's just the normal message handling loop.